ta/pkcs11/src/pkcs11_token.c

```diff
@@ -701,3 +701,118 @@ uint32_t entry_ck_session_info(struct pkcs11_client *client,
 
 	return PKCS11_CKR_OK;
 }
+
+uint32_t entry_ck_token_initialize(uint32_t ptypes, TEE_Param *params)
+{
+        const uint32_t exp_pt = TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INOUT,
+						TEE_PARAM_TYPE_NONE,
+						TEE_PARAM_TYPE_NONE,
+						TEE_PARAM_TYPE_NONE);
+	char label[PKCS11_TOKEN_LABEL_SIZE + 1] = { 0 };
+	struct pkcs11_client *client = NULL;
+	struct pkcs11_session *sess = NULL;
+	enum pkcs11_rc rc = PKCS11_CKR_OK;
+	struct serialargs ctrlargs = { };
+	struct ck_token *token = NULL;
+	TEE_Param *ctrl = params;
+	uint32_t token_id = 0;
+	uint32_t pin_size = 0;
+	void *pin = NULL;
+
+	if (ptypes != exp_pt)
+		return PKCS11_CKR_ARGUMENTS_BAD;
+
+	serialargs_init(&ctrlargs, ctrl->memref.buffer, ctrl->memref.size);
+
+	rc = serialargs_get(&ctrlargs, &token_id, sizeof(uint32_t));
+	if (rc)
+		return rc;
+
+	rc = serialargs_get(&ctrlargs, &pin_size, sizeof(uint32_t));
+	if (rc)
+		return rc;
+
+	if (pin_size < PKCS11_TOKEN_PIN_SIZE_MIN ||
+	    pin_size > PKCS11_TOKEN_PIN_SIZE_MAX)
+		return PKCS11_CKR_PIN_LEN_RANGE;
+
+	rc = serialargs_get(&ctrlargs, &label, PKCS11_TOKEN_LABEL_SIZE);
+	if (rc)
+		return rc;
+
+	rc = serialargs_get_ptr(&ctrlargs, &pin, pin_size);
+	if (rc)
+		return rc;
+
+	if (serialargs_remaining_bytes(&ctrlargs))
+		return PKCS11_CKR_ARGUMENTS_BAD;
+
+	token = get_token(token_id);
+	if (!token)
+		return PKCS11_CKR_SLOT_ID_INVALID;
+
+	if (token->db_main->flags & PKCS11_CKFT_SO_PIN_LOCKED) {
+		IMSG("Token %"PRIu32": SO PIN locked", token_id);
+		return PKCS11_CKR_PIN_LOCKED;
+	}
+
+	/* Check  there's no open session on this token */
+	TAILQ_FOREACH(client, &pkcs11_client_list, link)
+		TAILQ_FOREACH(sess, &client->session_list, link)
+			if (sess->token == token)
+				return PKCS11_CKR_SESSION_EXISTS;
+
+	if (!token->db_main->so_pin_salt) {
+		rc = hash_pin(PKCS11_CKU_SO, pin, pin_size,
+			      &token->db_main->so_pin_salt,
+			      token->db_main->so_pin_hash);
+		if (rc)
+			return rc;
+
+		update_persistent_db(token);
```

Comment:
can't this operation fail?

Commit "ta: pkcs11: helper to update token persistent database" should be placed before this commit in the series.

Reply:
No it can't fail, the TA panics instead since continueing from here could be a security problem.

Reply:
true, sorry. I made this comment before looking at the implementation :-1: